mlx/linalg.cpp

```diff
@@ -339,4 +339,40 @@ array cholesky_inv(
   }
 }
 
+std::pair<array, array> eigvalsh(
+    const array& a,
+    bool upper /* = true */,
+    bool compute_vectors /* = false */,
+    StreamOrDevice s /* = {} */) {
+  if (a.dtype() != float32) {
+    std::ostringstream msg;
+    msg << "[linalg::eigvalsh] Arrays must be type float32. Received array "
+        << "with type " << a.dtype() << ".";
+    throw std::invalid_argument(msg.str());
+  }
+
+  if (a.ndim() < 2) {
+    std::ostringstream msg;
+    msg << "[linalg::eigvalsh] Arrays must have >= 2 dimensions. Received array "
+           "with "
+        << a.ndim() << " dimensions.";
+    throw std::invalid_argument(msg.str());
+  }
+
+  if (a.shape(-1) != a.shape(-2)) {
+    throw std::invalid_argument(
+        "[linalg::eigvalsh] Eigenvalues are only defined for square matrices.");
+  }
+
+  std::vector<int> out_shape(a.shape().begin(), a.shape().end() - 1);
+  out_shape.back() = a.shape(-1);
+
+  auto out = array::make_arrays(
+      {out_shape, compute_vectors ? a.shape() : std::vector<int>()},
```

Comment:
I don't think you need to pass an empty second shape here if `compute_vectors=false`.
Instead, you could pass just one output if you only need eigenvalues and check `outputs.size()` in `Eigvalsh::eval`.